socks5.py

```diff
@@ -220,57 +376,131 @@ def read_addrport(self, address_type, sockfile):
         port, = readstruct(sockfile, "!H")
         return address, port
 
-    def tcp_loop(self, sock1, sock2):
+    def tcp_loop(self, csock, ssock):
+        global inbound_traffic
+        global outbound_traffic
         while True:
-            # wait until client or remote is available for read
-            r, _, _ = select([sock1, sock2], [], [], IDLE_TIMEOUT)
+            r, _, _ = select([csock, ssock], [], [], IDLE_TIMEOUT)
             if not r:
                 raise socket.timeout()
 
-            if sock1 in r:
-                data = sock1.recv(4096)
+            if csock in r:
+                data = csock.recv(4096)
                 if not data:
                     break
-                sock2.sendall(data)
+                ssock.sendall(data)
+                with traffic_lock:
+                    outbound_traffic += len(data)
 
-            if sock2 in r:
-                data = sock2.recv(4096)
+            if ssock in r:
+                data = ssock.recv(4096)
                 if not data:
                     break
-                sock1.sendall(data)
+                csock.sendall(data)
+                with traffic_lock:
+                    inbound_traffic += len(data)
+
 
     def handle_connect(self, address, port):
         log_tag = '%s:%s -> %s:%s' % (self.client_address + (address, port))
-        try:
-            remote = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            if CONNECT_HOST:
-                remote.bind((CONNECT_HOST, 0))
-            remote.connect((address, port))
-            logging.info('%s: connected', log_tag)
-        except Exception as err:
-            logging.error('%s: connect error %s', log_tag, err)
-            self.send_reply(self.STATUS_ECONNREFUSED)
-            self.server.close_request(self.request)
-            return
 
-        self.send_reply(self.STATUS_SUCCEEDED)
+        if isinstance(address, dict):
+            ipv4 = address.get('ipv4')
+            ipv6 = address.get('ipv6')
+        else:
+            if ':' in address:
+                ipv6 = address
+                ipv4 = None
+            else:
+                ipv4 = address
+                ipv6 = None
+
+        if (ipv6 is not None and "ipv6" in CONNECT_HOST and CONNECT_HOST["ipv6"]) and \
+        (ipv4 is not None and "ipv4" in CONNECT_HOST and CONNECT_HOST["ipv4"]):
+            with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+                if ipv6 is not None:
+                    ipv6_future = executor.submit(self.connect_to_address, ipv6, port, socket.AF_INET6, log_tag, timeout=5)
+                if ipv4 is not None:
+                    # Start IPv4 connection after 50ms delay
+                    ipv4_future = executor.submit(self.delayed_connect, ipv4, port, socket.AF_INET, 0.05, log_tag)
+
+                while True:
+                    if ipv6 is not None and ipv6_future.done():
```

Comment:
Can we use `concurrent.futures.wait(return_when=FIRST_COMPLETED)` instead of this loop?